server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -427,13 +438,39 @@ private static Query tryRewriteLongSort(SearchContext searchContext, IndexReader
         return rewrittenQuery;
     }
 
-    // Restore fieldsDocs to remove the first _score sort
-    // updating in place without creating new FieldDoc objects
+    /**
+     * Creates a sorter of {@link LeafReaderContext} that orders leaves depending on the minimum
+     * value and the sort order of the provided <code>sortField</code>.
+     */
+    static CheckedConsumer<List<LeafReaderContext>, IOException> createLeafSorter(SortField sortField) {
+        return leaves -> {
+            long[] minValues = new long[leaves.size()];
+            long missingValue = (long) sortField.getMissingValue();
+            for (LeafReaderContext ctx : leaves) {
+                PointValues values = ctx.reader().getPointValues(sortField.getField());
+                if (values == null) {
+                    minValues[ctx.ord] = (long) sortField.getMissingValue();
+                } else {
+                    byte[] minValue = values.getMinPackedValue();
+                    minValues[ctx.ord] = minValue == null ? missingValue : LongPoint.decodeDimension(minValue, 0);
+                }
+            }
+            Comparator<LeafReaderContext> comparator = Comparator.comparingLong(l -> minValues[l.ord]);
+            if (sortField.getReverse()) {
```

Comment:
I would think in the case of the reverse sort, we should look at the `maxValues` not `minValues`?